src/armscan_env/wrapper.py

```diff
@@ -128,26 +164,116 @@ def __getattr__(self, item: str) -> Any:
-    :param reward_metric: the reward metric to use
-    :param termination_criterion: the termination criterion to use
-    :param slice_shape: the shape of the slice
-    :param max_episode_len: the maximum episode length
-    :param rotation_bounds: the bounds for the angles
-    :param translation_bounds: the bounds for the translations
-    :param render_mode_train: the render mode to use for training environments
-    :param render_mode_test: the render mode to use for test environments
-    :param render_mode_watch: the render mode to use for environments that are used to watch agent performance
-    :param venv_type: the type of vectorized environment to use
-    :param seed: the seed to use
-    :param n_stack: the number of observations to stack in a single observation
-    :param project_to_x_translation: constrains the action space to only x translation
-    :param remove_rotation_actions: removes the rotation actions from the action space
-    :param make_kwargs: additional keyword arguments to pass to the environment creation function
+class AddObservationsWrapper(Wrapper, ABC):
+    """When implementing it, make sure that additional_obs_space is available
+    before super().__init__(env) is called.
     """
 
+    def __init__(self, env: LabelmapEnv | Env, additional_obs: Observation):
+        super().__init__(env)
+        self.additional_obs = additional_obs
+        if isinstance(self.env.observation_space, Box) and isinstance(
+            self.additional_obs_space,
+            Box,
+        ):
+            self.observation_space = ConcatenatedArrayObservation.concatenate_boxes(
+                [self.env.observation_space, self.additional_obs_space],
+            )
+        else:
+            raise ValueError(
+                f"Observation spaces are not of type Box: {type(self.env.observation_space)}, {type(self.additional_obs_space)}",
+            )
+
+    @property
+    @abstractmethod
+    def additional_obs_space(self) -> gym.spaces:
+        pass
+
+    @abstractmethod
+    def get_additional_obs(
+        self,
+    ) -> np.ndarray:
+        pass
+
+    def observation(
+        self,
+        observation: np.ndarray,
+    ) -> np.ndarray:
+        additional_obs = self.get_additional_obs()
+        try:
+            full_obs = np.concatenate([observation, additional_obs])
+        except ValueError:
+            raise ValueError(
+                f"Observation spaces are not of type Box: {type(observation)}, {type(additional_obs)}",
+            ) from None
+        return full_obs
+
+
+class AddRewardDetailsWrapper(AddObservationsWrapper):
+    @property
+    def additional_obs_space(self) -> gym.spaces:
+        return self._additional_obs_space
+
+    def __init__(
+        self,
+        env: LabelmapEnv | Env[ObsType, ActType],
+        num_steps_to_observe: int | None = None,
+        additional_obs: Observation = ActionRewardObservation().to_array_observation(),
```

Comment:
mutable defaults are evil. In the long run we need an observation registry and proper constructors. For now, pls just add a todo about adding that registry

Reply:
I feel like this shouldn't be passed here at all. This wrapper is supposed to add reward details, and in fact no other additional_obs can be passed (it would break your code below). Would also solve the registry problem, no need for a todo then

Reply:
But then how do I set the additional observation space? since it depends on the action_space.shape

Reply:
In the call to super init. No need to let the user pass it